src/compiler/checker.ts

```diff
@@ -14092,7 +14106,10 @@ namespace ts {
                 if (assignedType.flags & TypeFlags.Never) {
                     return assignedType;
                 }
-                const reducedType = filterType(declaredType, t => typeMaybeAssignableTo(assignedType, t));
+                let reducedType = filterType(declaredType, t => typeMaybeAssignableTo(assignedType, t));
+                if (assignedType.flags & (TypeFlags.FreshLiteral | TypeFlags.Literal)) {
+                    reducedType = mapType(reducedType, getFreshTypeOfLiteralType); // Ensure that if the assignment is a fresh type, that we narrow to fresh types

```

Comment:
why didn't we have to do this before, for string and number literal types?

Reply:
_We probably should have been_ but as it turns out people narrow `boolean` made via `!!expr || !expr` a ton more than narrowing `x === "a" | x === "b"`, so it's a lot more obvious an issue with booleans.

Reply:
And as I said in person, this is going to break one RWC project along the lines of this:
```ts
type RuleSeverity = "warning" | "error" | "off";
interface IOptions {
    ruleSeverity: RuleSeverity;
}

function parseRuleOptions(): Partial<IOptions> {
    let defaultRuleSeverity: RuleSeverity = "error";
    let ruleSeverity = defaultRuleSeverity;

    return {
        ruleSeverity,
    };
}
```
previously this worked because `defaultRuleSeverity` was narrowed to the nonfresh `"error"` when it was used to initialize `ruleSeverity`, whereas now we choose the fresh `"error"` which widens (and triggers an error in the return type below). So this is a (probably small) break.

Reply:
FWIW the individual unit type only behaves differently because it skips the narrowing codepath. Had I realized as much I probably would've altered the unit type behavior, too. In any case, removing this bit if we don't like it is easy; but it's going to make boolean literals a tad more breaky.